src/impl_methods.rs

```diff
@@ -1329,6 +1329,58 @@ impl<A, S, D> ArrayBase<S, D> where S: Data<Elem=A>, D: Dimension
         self.strides.slice_mut().swap(ax, bx);
     }
 
+    /// Permute the axes.
+    ///
+    /// This does not move any data, it just adjusts the array’s dimensions
+    /// and strides.
+    ///
+    /// *i* in the *j*-th place in the axes sequence means `self`'s *i*-th axis
+    /// becomes `self.permuted_axes()`'s *j*-th axis
+    ///
+    /// **Panics** if any of the axes are out of bounds, if an axis is missing,
+    /// or if an axis is repeated more than once.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ndarray::{arr2, Array3};
+    ///
+    /// let a = arr2(&[[0, 1], [2, 3]]);
+    /// assert_eq!(a.view().permuted_axes([1, 0]), a.t());
+    ///
+    /// let b = Array3::<u8>::zeros((1, 2, 3));
+    /// assert_eq!(b.permuted_axes([1, 0, 2]).shape(), &[2, 1, 3]);
+    /// ```
+    pub fn permuted_axes<T>(mut self, axes: T) -> ArrayBase<S, D>
+    where
+        T: IntoDimension<Dim = D>,
+    {
+        let axes = axes.into_dimension();
+        // Ensure that each axis is used exactly once.
+        let mut usage_counts = D::zero_index_with_ndim(self.ndim());
+        for axis in axes.slice() {
+            usage_counts[*axis] += 1;
+        }
+        for count in usage_counts.slice() {
+            assert_eq!(*count, 1, "each axis must be listed exactly once");
+        }
+        // Determine the new shape and strides.
+        let mut new_dim = D::zero_index_with_ndim(self.ndim());
```

Comment:
Instead of creating a new `D` here for `new_dim`, we can re-purpose `usage_counts`.

Reply:
Good idea! This brings the speed for `IxDyn` closer to that of the in-place implementation without noticeably changing the fixed-dimension cases. Now, the largest difference from the in-place implementation is ~40 ns/iter.

Reply:
Nice